File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/finder/RSSFinder.java
package lan.dk.podcastserver.manager.worker.finder;

import io.vavr.control.Option;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.service.ImageService;
import lan.dk.podcastserver.service.JdomService;
import lombok.RequiredArgsConstructor;
import org.hibernate.validator.constraints.NotEmpty;
import org.jdom2.Element;
import org.springframework.stereotype.Service;

import java.util.Objects;

/**
 * Created by kevin on 22/02/15
 */
@Service("RSSFinder")
@RequiredArgsConstructor
public class RSSFinder implements Finder {

    private static final String CHANNEL = "channel";
    private static final String TITLE = "title";
    private static final String DESCRIPTION = "description";
    private static final String IMAGE = "image";
    private static final String URL = "url";
    private static final String HREF = "href";

    private final JdomService jdomService;
    private final ImageService imageService;

    @Override
    public Podcast find(String url) {
        return jdomService
                .parse(url)
                .map(x -> x.getRootElement().getChild(CHANNEL))
                .filter(Objects::nonNull)
                .map(e -> this.xmlToPodcast(e, url))
                .getOrElse(Podcast.DEFAULT_PODCAST);
    }

    private Podcast xmlToPodcast(Element element, String url) {
        return Podcast.builder()
                .type("RSS")
                .url(url)
                .title(element.getChildText(TITLE))
                .description(element.getChildText(DESCRIPTION))
                .cover(coverUrlOf(element))
                .build();
    }

    private Cover coverUrlOf(Element channelElement) {
        return getRssImage(channelElement)
                .orElse(getItunesImage(channelElement))
                .map(imageService::getCoverFromURL)
                .getOrElse(Cover.DEFAULT_COVER);
    }

    private Option<String> getItunesImage(Element channelElement) {
        return Option.of(channelElement.getChild(IMAGE, JdomService.ITUNES_NAMESPACE))
                .map(itunesImage -> itunesImage.getAttributeValue(HREF));
    }

    private Option<String> getRssImage(Element channelElement) {
        return Option.of(channelElement.getChild(IMAGE))
                .map(rssImage -> rssImage.getChildText(URL));
    }

    public Integer compatibility(@NotEmpty String url) {
        return Integer.MAX_VALUE-1;
    }

}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/rss/RSSFinder.kt
package com.github.davinkevin.podcastserver.manager.worker.rss

import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.find.FindCoverInformation
import com.github.davinkevin.podcastserver.find.FindPodcastInformation
import com.github.davinkevin.podcastserver.manager.worker.Finder
import com.github.davinkevin.podcastserver.service.CoverInformation
import org.jdom2.Element
import org.jdom2.Namespace
import org.jdom2.input.SAXBuilder
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToMono
import reactor.core.publisher.Mono
import reactor.core.publisher.toMono
import java.io.ByteArrayInputStream
import java.net.URI

/**
 * Created by kevin on 22/02/15
 */
@Service
class RSSFinder(
        val imageService: ImageService,
        val wcb: WebClient.Builder
) : Finder {

    private val itunesNS = Namespace.getNamespace("itunes", "http://www.itunes.com/dtds/podcast-1.0.dtd")!!

    override fun find(url: String): Podcast {
        TODO("not required anymore") //To change body of created functions use File | Settings | File Templates.
    }

    override fun findInformation(url: String): Mono<FindPodcastInformation> = wcb
            .baseUrl(url)
            .build()
            .get()
            .retrieve()
            .bodyToMono<String>()
            .map { SAXBuilder().build(ByteArrayInputStream(it.toByteArray(Charsets.UTF_8))) }
            .map { it.rootElement.getChild("channel") }
            .filter { it != null }
            .map { xmlToPodcast(it, url) }
            .onErrorResume {
                FindPodcastInformation(title = "", url = URI(url), type = "RSS", cover = null, description = "")
                        .toMono()
            }

    private fun xmlToPodcast(element: Element, anUrl: String) =
            FindPodcastInformation(
                type = "RSS",
                url = URI(anUrl),
                title = element.getChildText("title"),
                description = element.getChildText("description"),
                cover = coverUrlOf(element)?.toFindCover()
            )

    private fun coverUrlOf(channelElement: Element): CoverInformation? {
        val rss = channelElement.getChild("image")?.getChildText("url")
        val itunes = channelElement.getChild("image", itunesNS)?.getAttributeValue("href")

        val url = rss ?: itunes ?: return null

        return imageService.fetchCoverInformation(url)
    }

    override fun compatibility(url: String?) = Integer.MAX_VALUE - 1
}

private fun CoverInformation.toFindCover() = FindCoverInformation(height, width, url)
